include/up-cpp/datamodel/builder/UMessage.h

```diff
@@ -0,0 +1,258 @@
+
+	/// @brief Pre-populates a message builder with the attributes of an
+	///        RPC "request" type message.
+	///
+	/// @param method UUri where the RPC request will be serviced.
+	/// @param source UUri where responses will be returned to.
+	/// @param priority Priority of built request messages.
+	/// @param ttl Amount of time from the call to build() that a message
+	///            will still be considered valid. Must be greater than 0
+	///            and fit within a uint32_t.
+	///
+	/// @throws InvalidUUri if URI fails validation.
+	/// @throws std::out_of_range if (ttl <= 0) or (ttl > uint32_t max).
+	/// @returns UMessageBuilder configured to build a "request" message
+	static UMessageBuilder request(v1::UUri&& method, v1::UUri&& source,
+	                               v1::UPriority priority,
+	                               std::chrono::milliseconds ttl);
+
+	/// @brief Pre-populates a message builder with the attributes of an
+	///        RPC "response" type message.
+	///
+	/// @note An alternate response(UMessage) interface is also provided
+	///       that will, in many cases, be simpler to use than this
+	///       interface.
+	///
+	/// @param sink UUri where the RPC will be delivered.
+	/// @param request_id ID from the originating request.
+	/// @param priority Priority from the originating request.
+	/// @param method UUri where the RPC request was serviced.
+	///
+	/// @throws InvalidUuid if ID fails validation.
+	/// @throws InvalidUUri if URI fails validation.
+	/// @returns UMessageBuilder configured to build a "response" message
+	static UMessageBuilder response(v1::UUri&& sink, v1::UUID&& request_id,
+	                                v1::UPriority priority, v1::UUri&& method);
+
+	/// @brief Pre-populates a message builder with the attributes of an
+	///        RPC "response" type message based on the initiating request
+	///        message.
+	///
+	/// @param request The request UMesssage to generate a response to.
+	///
+	/// @throws InvalidUuid if ID fails validation.
+	/// @throws InvalidUUri if URI fails validation.
+	/// @returns UMessageBuilder configured to build a "response" message
+	static UMessageBuilder response(const v1::UMessage& request);
+
+	/// @brief Specify an ID to use for the message in place of the
+	///        auto-generated ID.
+	///
+	/// Allows for a fixed, known ID to be set as the message ID. Normally,
+	/// a unique message ID would be generated each time build() is called,
+	/// but a builder where withId() has been called will always produce the ID
+	/// used in that call.
+	///
+	/// @note In general, use of this method represents a special case. It
+	///       does not need to be used in most scenarios as a unique
+	///       message ID will be generated automatically for every built
+	///       message.
+	///
+	/// @post This UMessageBuilder will only generate messages with their
+	///       ID set to the value passed to this method.
```

Comment:
Could maybe leave a further blurb warning about this method, given that the UUID contains not only the `rand_b` portion for uniquely identifying the uE, but also contains the timestamp & counter.

Just trying to think of more flashing neon signs to put above this function say -- "hey you probably didn't mean to use this!"

Reply:
I'm thinking it might be safer to just remove. The message builder will be inserting valid ones automatically, and if there is a need to replace them (e.g. for testing or for message playback), that can be done after calling build()